doc/design/api.md

```diff
@@ -0,0 +1,277 @@
+# PaddlePaddle Design Doc
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test(cost, ...)
```

Comment:
I don't totally understand why "Model should be bound to a device(a GPU card, or all CPUs)". I think parameters should be place by paddle automatically. Paddle should try it's best to place all parameters on GPU. The only case it fails to do so is when a layer don't have GPU implementation. So parameter will be allocated on CPU memory. And input tensor for that layer will be copied from GPU to CPU. output tensor will be copied from CPU to GPU.

It will be same for multi-GPU case, model is replicated n times where n is number of GPUs. So when initializing `Evaluator` and `Gradient Machine`, constructor parameters need to include GPU device ids as a list, `len(list)` will be how many model replicas we have.